### DISCUSSION TITLE
DisposeNotRequiredAttribute: Add way to mark IDisposable classes as not needing to be disposed

### DISCUSSION THREAD

**Discussion by Grauenwolf** [dotnet/csharplang#6222] [Language Ideas] [answered]
Many classes implement `IDisposable` for legacy reasons, but don't actually need to be disposed. This could be because they inherit from `Component` or because of a design flaw that was never corrected.

When enabling Warning Wave 5 in the compiler (`<AnalysisLevel>5.0-all</AnalysisLevel>`), we see the a lot of CA2000 warnings that could be classified as false positives.

Proposal:

1. Introduce a `DisposeNotRequired` attribute.
2. Mark the attribute as not-inheritable. Sub-classes may need to be disposed.
3. Change the compiler to not issue a warning if the class has the `DisposeNotRequired` attribute.

**Comment by 333fred** [comment, marked as answer]
I'm unsure why I can't transfer the issue directly, but this should be raised on the https://github.com/dotnet/roslyn-analyzers repo, not on csharplang.

**Reply by Grauenwolf** [reply]
From what I understand, this is now part of the C# Compiler itself.

**Reply by 333fred** [reply]
Nope. It is part of the SDK, and the implementation lives in the repo I mentioned above. Warnings that come from roslyn itself always start with a `CS` prefix (for C# errors).

**Reply by Grauenwolf** [reply]
Moved to https://github.com/dotnet/roslyn-analyzers/issues/6028

**Reply by Joe4evr** [reply]
> From what I understand, this is now part of the C# Compiler itself.

Not to mention this repo is *not* about the compiler either. 🙃
